lib/pengine/tests/native/native_find_rsc_test.c

```diff
@@ -0,0 +1,683 @@
+
+    /* Get references to several resources we use frequently. */
+    for (GList *iter = data_set->resources; iter != NULL; iter = iter->next) {
+        pe_resource_t *rsc = (pe_resource_t *) iter->data;
+
+        if (strcmp(rsc->id, "exim-group") == 0) {
+            exim_group = rsc;
+        } else if (strcmp(rsc->id, "httpd-bundle") == 0) {
+            httpd_bundle = rsc;
+        } else if (strcmp(rsc->id, "inactive-clone") == 0) {
+            inactive_clone = rsc;
+        } else if (strcmp(rsc->id, "inactive-group") == 0) {
+            inactive_group = rsc;
+        } else if (strcmp(rsc->id, "mysql-clone-group") == 0) {
+            mysql_clone_group = rsc;
+        } else if (strcmp(rsc->id, "promotable-clone") == 0) {
+            promotable_clone = rsc;
+        }
+    }
+
+    return 0;
+}
+
+static int
+teardown(void **state) {
+    pe_free_working_set(data_set);
+
+    return 0;
+}
+
+static void
+bad_args(void **state) {
+    pe_resource_t *rsc = (pe_resource_t *) g_list_first(data_set->resources)->data;
+    char *id = rsc->id;
+    char *name = NULL;
+
+    assert_non_null(rsc);
+
+    assert_null(native_find_rsc(NULL, "dummy", NULL, 0));
+    assert_null(native_find_rsc(rsc, NULL, NULL, 0));
+
+    /* No resources exist with these names. */
+    name = crm_strdup_printf("%sX", rsc->id);
+    assert_null(native_find_rsc(rsc, name, NULL, 0));
+    free(name);
+
+    name = crm_strdup_printf("x%s", rsc->id);
+    assert_null(native_find_rsc(rsc, name, NULL, 0));
+    free(name);
+
+    name = g_ascii_strup(rsc->id, -1);
+    assert_null(native_find_rsc(rsc, name, NULL, 0));
+    g_free(name);
+
+    /* Fails because resource ID is NULL. */
+    rsc->id = NULL;
+    assert_null(native_find_rsc(rsc, id, NULL, 0));
+    rsc->id = id;
+}
+
+static void
+primitive_rsc(void **state) {
+    pe_resource_t *dummy = NULL;
+
+    /* Find the "dummy" resource, which is the only one with that ID in the set. */
+    for (GList *iter = data_set->resources; iter != NULL; iter = iter->next) {
+        pe_resource_t *rsc = (pe_resource_t *) iter->data;
+
+        if (strcmp(rsc->id, "dummy") == 0) {
+            dummy = rsc;
+            break;
+        }
+    }
+
+    assert_non_null(dummy);
+
+    /* Passes because NULL was passed for node, regardless of flags. */
+    assert_ptr_equal(dummy, native_find_rsc(dummy, "dummy", NULL, 0));
+    assert_ptr_equal(dummy, native_find_rsc(dummy, "dummy", NULL, pe_find_current));
+
+    /* Fails because resource is not a clone (nor cloned). */
+    assert_null(native_find_rsc(dummy, "dummy", NULL, pe_find_clone));
+    assert_null(native_find_rsc(dummy, "dummy", cluster02, pe_find_clone));
+
+    /* Fails because dummy is not running on cluster01, even with the right flags. */
+    assert_null(native_find_rsc(dummy, "dummy", cluster01, pe_find_current));
+
+    /* Fails because pe_find_current is required if a node is given. */
+    assert_null(native_find_rsc(dummy, "dummy", cluster02, 0));
```

Comment:
To clarify further, this fails only because we didn't call pcmk__schedule_actions(). If we had done that, then (in this case) dummy is staying on cluster02, so the allocated node would match. But since we only called cluster_status(), we don't have any allocated node information.

To put it another way -- we could call cluster_status(), run this test as is, then call pcmk__schedule_actions(), then repeat this test, and it would pass on the second call.

Alternatively, we could just call pcmk__schedule_actions() instead of cluster_status(), and be able to test current vs allocated nodes.

Reply:
I don't think it would be difficult to test both before and after running pcmk__schedule_actions, just verbose.  I think what we could do is pass some state variable to each test - say, a bool that's false if these should pass.  We'd run the tests, call pcmk__schedule_actions, and then run them again with that bool now set to true.  The verbose part would be conditionalizing all these tests that can later pass.